src/threadNames.cpp

```diff
@@ -0,0 +1,52 @@
+#include "threadNames.h"
+#include "profiler.h"
+
+
+jfieldID ThreadNames::_get_eetop(JNIEnv *env) {
+    MutexLocker ml(_cached_eetop_lock);
+    if (_cached_eetop == NULL) {
+        jclass threadClass = env->FindClass("java/lang/Thread");
+        if (threadClass != NULL) {
+            _cached_eetop = env->GetFieldID(threadClass, "eetop", "J");
+        }
+    }
+    return _cached_eetop;
+}
+
+void ThreadNames::update(jthread thread) {
+    if (!VMThread::available()) {
+        return;
+    }
+    JNIEnv *env = VM::jni();
+    jfieldID eetop = _get_eetop(env);
+    if (eetop == NULL) {
+        return;
+    }
+    VMThread *vm_thread = (VMThread *)(uintptr_t)env->GetLongField(thread, eetop);
+    jvmtiThreadInfo info;
+    jvmtiEnv *jvmti = VM::jvmti();
+    if (jvmti->GetThreadInfo(thread, &info) != 0 || info.name == NULL) {
+        return;
+    }
+    MutexLocker ml(_storage_lock);
```

Comment:
Narrowing the scope of the mutex:
```
{
    MutexLocker ml(_storage_lock);
    _storage[vm_thread->osThreadId()] = std::string(info.name);
}
```

Reply:
Forget about this language feature, will fix.